torch_geometric/nn/models/rbcd_attack.py

```diff
@@ -0,0 +1,755 @@
+                current_block, sorted=True, return_inverse=True)
+
+            if self.is_undirected_graph:
+                self.block_edge_index = self._linear_to_triu_idx(
+                    self.num_nodes, self.current_block)
+            else:
+                self.block_edge_index = self._linear_to_full_idx(
+                    self.num_nodes, self.current_block)
+
+            # Merge existing weights with new edge weights
+            block_edge_weight_prev = self.block_edge_weight[sorted_idx]
+            self.block_edge_weight = torch.full(self.current_block.shape,
+                                                self.coeffs['eps'],
+                                                device=self.device)
+            self.block_edge_weight[
+                unique_idx[:sorted_idx.size(0)]] = block_edge_weight_prev
+
+            if not self.is_undirected_graph:
+                self._filter_self_loops_in_block(with_weight=True)
+
+            if self.current_block.size(0) > budget:
+                return
+        raise RuntimeError('Sampling random block was not successful.'
+                           'Please decrease `budget`.')
+
+    def _sample_final_edges(self, x: Tensor, labels: Tensor, budget: int,
+                            idx_attack: Optional[Tensor] = None,
+                            **kwargs) -> Tuple[Tensor, Tensor]:
+        best_metric = float('-Inf')
+        block_edge_weight = self.block_edge_weight
+        block_edge_weight[block_edge_weight <= self.coeffs['eps']] = 0
+
+        for i in range(self.coeffs['max_final_samples']):
+            if i == 0:
+                # In first iteration employ top k heuristic instead of sampling
+                sampled_edges = torch.zeros_like(block_edge_weight)
+                sampled_edges[torch.topk(block_edge_weight,
+                                         budget).indices] = 1
+            else:
+                sampled_edges = torch.bernoulli(block_edge_weight).float()
+
+            if sampled_edges.sum() > budget:
+                # Allowed budget is exceeded
+                continue
+            self.block_edge_weight = sampled_edges
+
+            edge_index, edge_weight = self._get_modified_adj(
+                self.edge_index, self.edge_weight, self.block_edge_index,
+                self.block_edge_weight)
+            prediction = self._forward(x, edge_index, edge_weight, **kwargs)
+            metric = self.metric(prediction, labels, idx_attack)
+
+            # Save best sample
+            if metric > best_metric:
+                best_metric = metric
+                best_edge_weight = self.block_edge_weight.clone().cpu()
+
+        # Recover best sample
+        self.block_edge_weight = best_edge_weight.to(self.device)
+        flipped_edges = self.block_edge_index[:,
+                                              torch.where(best_edge_weight)[0]]
+
+        edge_index, edge_weight = self._get_modified_adj(
```

Comment:
[nit] The following code look better to me:
```python
edge_index, edge_weight = self._get_modified_adj(self.edge_index,
                                                 self.edge_weight,
                                                 self.block_edge_index,
                                                 self.block_edge_weight)
```

Reply:
Does hit the max number of characters